cmd/common/oci.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends/console"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	clioperator "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/cli"
+	ocioperator "github.com/inspektor-gadget/inspektor-gadget/pkg/operators/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumnTags []string) *cobra.Command {
+	runtimeGlobalParams := runtime.GlobalParamDescs().ToParams()
+	runtime.Init(runtimeGlobalParams)
+
+	ociParams := ocioperator.OciHandler.ParamDescs().ToParams()
+
+	// Add operator global flags
+	operatorsGlobalParamsCollection := operators.GlobalParamsCollection()
+
+	cmd := &cobra.Command{
+		Use:          "run-experimental",
+		Short:        "run gadget",
+		SilenceUsage: true, // do not print usage when there is an error
+		// We have to disable flag parsing in here to be able to handle certain
+		// flags more dynamically and have `--help` also react to those changes.
+		// Instead, we need to manually
+		// * call cmd.ParseFlags()
+		// * handle `--help` after changing the params dynamically
+		// * handle everything that could have been handled inside
+		//   `PersistentPreRun(E)` of a parent cmd, as the flags wouldn't have
+		//   been parsed there (e.g. --verbose)
+		DisableFlagParsing: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// Manually register the clioperator
+			clioperator.Register()
```

Comment:
To add more on this. I think we can continue using this approach for the CLI, but later on I'd like to see this used for all operators instead of relying only on the import. It'll help with https://github.com/inspektor-gadget/inspektor-gadget/issues/1992 and also with applications that need to use multiple operators but not enable all of them.

Reply:
> To add more on this. I think we can continue using this approach for the CLI, but later on I'd like to see this used for all operators instead of relying only on the import. It'll help with #1992 and also with applications that need to use multiple operators but not enable all of them.

Hmm, as the decision whether to use an operator or not happens at runtime (not at compile time), I think we should always import/register them by default. Which operator actually gets used then should be decided when the operators are iterated over (and by the operators themselves of course).

We probably also need to make the operators aware of the environment they're used in: for example, operators that only add fields for conversion of existing fields (timestamp => string) could be run only on the client side of things to avoid sending redundant information over the network (but this needs some further adjustments in DataSource handling and should be added at a later stage).

Reply:
> Hmm, as the decision whether to use an operator or not happens at runtime (not at compile time), I think we should always import/register them by default. Which operator actually gets used then should be decided when the operators are iterated over (and by the operators themselves of course).

I'm fine with always registering them. But there should be a way to disable them programmatically. It could happen that an application includes many of them but they should be enabled in specific cases. (I think this case of only enabling the cli operator when the binary needs to print something is only one example of multiple cases where this could happen).

Reply:
> > Hmm, as the decision whether to use an operator or not happens at runtime (not at compile time), I think we should always import/register them by default. Which operator actually gets used then should be decided when the operators are iterated over (and by the operators themselves of course).
> 
> I'm fine with always registering them. But there should be a way to disable them programmatically. It could happen that an application includes many of them but they should be enabled in specific cases. (I think this case of only enabling the cli operator when the binary needs to print something is only one example of multiple cases where this could happen).

Totally agree - do you want this to be implemented in this PR already, though?

Reply:
Nope, this can be handled later on!